generators/python-v2/ast/src/PythonFile.ts

```diff
@@ -44,47 +48,99 @@ export class PythonFile extends AstNode {
     }
 
     public write(writer: Writer): void {
+        const uniqueReferences = this.deduplicateReferences();
+
+        this.updateWriterRefNameOverrides({ writer, uniqueReferences });
+
         this.writeComments(writer);
-        this.writeImports(writer);
+        this.writeImports({ writer, uniqueReferences });
         this.statements.forEach((statement, idx) => {
             statement.write(writer);
             writer.newLine();
             if (idx < this.statements.length - 1) {
                 writer.newLine();
             }
         });
+
+        writer.unsetRefNameOverrides();
     }
 
     /*******************************
      * Helper Methods
      *******************************/
 
-    private writeComments(writer: Writer): void {
-        this.comments.forEach((comment) => {
-            comment.write(writer);
+    private updateWriterRefNameOverrides({
+        writer,
+        uniqueReferences
+    }: {
+        writer: Writer;
+        uniqueReferences: Map<string, { modulePath: ModulePath; references: Reference[]; referenceNames: Set<string> }>;
+    }): void {
+        const reservedNames = this.getReservedNames();
+        const references: Reference[] = Array.from(uniqueReferences.values()).flatMap(({ references }) => references);
+
+        // Build up a map of refs to their name overrides, keeping track of howmany times we've seen a name as we go.
+        const completeRefPathsToNameOverrides: Record<string, { name: string; isAlias: boolean }> = {};
+        const nameUsageCounts: Record<string, number> = {};
```

Comment:
Do we actually need to track the counts? We only care if it's used at all right? Could we just use `reservedNames` instead?

Reply:
We do for the current naming convention scheme, but once I adapt to match Vargas's proposal, we won't need to.